src/storage/read_pool_impl.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cell::RefCell;
+use std::time::Duration;
+
+use prometheus::local::*;
+
+use crate::pd::PdTask;
+use crate::server::readpool::{self, Builder, ReadPool};
+use crate::util::collections::HashMap;
+use crate::util::worker::FutureScheduler;
+
+use super::metrics::*;
+
+thread_local! {
+    static LOCAL_SCHED_HISTOGRAM_VEC: RefCell<LocalHistogramVec> =
+        RefCell::new(SCHED_HISTOGRAM_VEC.local());
+
+    static LOCAL_SCHED_PROCESSING_READ_HISTOGRAM_VEC: RefCell<LocalHistogramVec> =
+        RefCell::new(SCHED_PROCESSING_READ_HISTOGRAM_VEC.local());
+
+    static LOCAL_KV_COMMAND_KEYREAD_HISTOGRAM_VEC: RefCell<LocalHistogramVec> =
+        RefCell::new(KV_COMMAND_KEYREAD_HISTOGRAM_VEC.local());
+
+    static LOCAL_KV_COMMAND_COUNTER_VEC: RefCell<LocalIntCounterVec> =
+        RefCell::new(KV_COMMAND_COUNTER_VEC.local());
+
+    static LOCAL_SCHED_COMMANDS_PRI_COUNTER_VEC: RefCell<LocalIntCounterVec> =
+        RefCell::new(SCHED_COMMANDS_PRI_COUNTER_VEC.local());
+
+    static LOCAL_KV_COMMAND_SCAN_DETAILS: RefCell<LocalIntCounterVec> =
+        RefCell::new(KV_COMMAND_SCAN_DETAILS.local());
+
+    static LOCAL_PD_SENDER: RefCell<Option<FutureScheduler<PdTask>>> =
+        RefCell::new(None);
+
+    static LOCAL_READ_FLOW_STATS: RefCell<HashMap<u64, crate::storage::FlowStatistics>> =
+        RefCell::new(HashMap::default());
+}
+
+pub struct ReadPoolImpl;
+
+impl ReadPoolImpl {
+    #[inline]
+    pub fn build_read_pool(
+        config: &readpool::Config,
+        pd_sender: FutureScheduler<PdTask>,
+        name_prefix: &str,
+    ) -> ReadPool {
+        let pd_sender2 = pd_sender.clone();
+
+        Builder::from_config(config)
+            .name_prefix(name_prefix)
+            .on_tick(move || ReadPoolImpl::thread_local_flush(&pd_sender))
+            .before_stop(move || ReadPoolImpl::thread_local_flush(&pd_sender2))
+            .build()
+    }
+
+    #[inline]
+    fn thread_local_flush(pd_sender: &FutureScheduler<PdTask>) {
```

Comment:
I think using thread_local everywhere is long and redundant. If we want to represent thread_local, mostly, we can use `tls` instead.

Reply:
Yeah, tls looks to be a good name!